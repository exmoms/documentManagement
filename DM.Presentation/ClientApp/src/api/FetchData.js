/**
 * @file this file includes functions that fetch data from API
 */

// Generic Delete Method
export async function deleteFromApi(url) {
  let res = await fetch(url, {
    method: "DELETE",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  });
  return res;
}

// Get MetaData Models
export async function fetchData(url) {
  let object = {};
  /*
   * send the request through a CORS proxy
   * https://stackoverflow.com/questions/43262121/trying-to-use-fetch-and-pass-in-mode-no-cors/43268098#43268098
   */
  await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  })
    .then((response) => response.json())
    .then(
      (data) => (object = data),
      (error) => console.log(error)
    );
  return object;
}
