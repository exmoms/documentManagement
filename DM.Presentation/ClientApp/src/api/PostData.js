/**
 * @file This file includes functions that help to post data to a specific API
 * @version 1.0
 */

/**
 * Post new document data.
 * Upload form-data, scanned copy of doc and attachments.
 * @author Youssef Shaaban <y.shaaban@lit-co.net>
 * @param {*} documentData
 * @param {*} images
 * @param {*} attachments
 * @param {*} url
 * @param {*} option its values either "Add" for add new document version or "Update" to update exsiting document Version
 */
export async function postDocumentData(
  documentData,
  scannedPages,
  attachments,
  url = "",
  option
) {
  var formData = new FormData();

  if (process.env.NODE_ENV === "development") {
    console.log(documentData);
    console.log(attachments);
    console.log(scannedPages);
  }

  switch (option) {
    case "Add":
      formData.append("Document", JSON.stringify(documentData));
      break;
    case "Update":
      formData.append("Version", JSON.stringify(documentData));
      break;
    default:
  }

  attachments.forEach((attachment) => {
    formData.append("Attachments", attachment);
  });
  scannedPages.forEach((scannedPage) => {
    formData.append("Scans", scannedPage);
  });

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    headers: {
      Accept: "application/json",
      // "Content-Type": "application/json",
    },
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Attachment Data.
 * Upload attachment json data + attachment file  .
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} payload
 * @param {*} attachmentFile
 * @param {*} url
 */
export async function postAttachmentData(payload, attachmentFile, url = "") {
  var formData = new FormData();

  formData.append("Attachment", JSON.stringify(payload));
  formData.append("FileAttachment", attachmentFile);

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Request To API
 * Send post request to api with  payload  and return a response.
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} url The api url
 * @param {*} payload js object which will be sent as request body.
 */
export async function postDataToAPI(url, payload = null, method = "POST") {

  const response = await fetch(url, {
    method: method,
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
  return response;
}
