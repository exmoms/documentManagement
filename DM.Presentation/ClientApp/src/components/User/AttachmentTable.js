/**
 * @file Contains AttachmentTable component implementaion
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @version 1.0
 */

import React, { Component, forwardRef } from "react";
import MaterialTable from "material-table";
import { Notification } from "../Admin/Notifications";
import { postDataToAPI, postAttachmentData } from "../../api/PostData";

import AddBox from "@material-ui/icons/AddBox";
import ArrowDownward from "@material-ui/icons/ArrowDownward";
import Check from "@material-ui/icons/Check";
import ChevronLeft from "@material-ui/icons/ChevronLeft";
import ChevronRight from "@material-ui/icons/ChevronRight";
import Clear from "@material-ui/icons/Clear";
import DeleteOutline from "@material-ui/icons/DeleteOutline";
import Edit from "@material-ui/icons/Edit";
import FilterList from "@material-ui/icons/FilterList";
import FirstPage from "@material-ui/icons/FirstPage";
import LastPage from "@material-ui/icons/LastPage";
import Remove from "@material-ui/icons/Remove";
import SaveAlt from "@material-ui/icons/SaveAlt";
import Search from "@material-ui/icons/Search";
import DeleteIcon from "@material-ui/icons/Delete";
import GetAppIcon from "@material-ui/icons/GetApp";
import UpdateIcon from "@material-ui/icons/Update";
import ViewColumn from "@material-ui/icons/ViewColumn";
import { withTranslation } from "react-i18next";

const tableIcons = {
  Add: forwardRef((props, ref) => <AddBox {...props} ref={ref} />),
  Check: forwardRef((props, ref) => <Check {...props} ref={ref} />),
  Clear: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
  Delete: forwardRef((props, ref) => <DeleteOutline {...props} ref={ref} />),
  DetailPanel: forwardRef((props, ref) => (
    <ChevronRight {...props} ref={ref} />
  )),
  Edit: forwardRef((props, ref) => <Edit {...props} ref={ref} />),
  Export: forwardRef((props, ref) => <SaveAlt {...props} ref={ref} />),
  Filter: forwardRef((props, ref) => <FilterList {...props} ref={ref} />),
  FirstPage: forwardRef((props, ref) => <FirstPage {...props} ref={ref} />),
  LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} />),
  NextPage: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
  PreviousPage: forwardRef((props, ref) => (
    <ChevronLeft {...props} ref={ref} />
  )),
  ResetSearch: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
  Search: forwardRef((props, ref) => <Search {...props} ref={ref} />),
  SortArrow: forwardRef((props, ref) => <ArrowDownward {...props} ref={ref} />),
  ThirdStateCheck: forwardRef((props, ref) => <Remove {...props} ref={ref} />),
  ViewColumn: forwardRef((props, ref) => <ViewColumn {...props} ref={ref} />),
};

class AttachmentTable extends Component {
  inputRefs = new Set();
  constructor(props) {
    super(props);
    this.state = {
      openNotifiction: false,
      error: false,
      errorMessage: [],
    };
    this.updateParentAttachments = this.props.updateAttachments;
  }

  t = this.props["t"];

  deleteRow = (rowData) => {
    new Promise((resolve, reject) => {
      setTimeout(() => {
        {
          let attachments = this.props.attachments;
          const index = attachments.indexOf(rowData);
          attachments.splice(index, 1);
          this.setState(
            {
              openNotifiction: true,
            },
            this.updateParentAttachments(attachments)
          );
        }
        resolve();
      }, 1000);
    });
  };

  deleteAttachment = async (rowData) => {
    let response = await postDataToAPI(
      `/api/Document/DeleteAttachments?id=${rowData.id}`,
      null
    );

    if (response.ok) {
      this.deleteRow(rowData);
    } else {
      let json = await response.json();
      this.setState({
        openNotifiction: true,
        error: true,
        errorMessage: json.error,
      });
    }
  };

  updateAttachmentcOnClickHandler = (attachment) => {
    var temp = [...this.inputRefs];
    var index = temp.findIndex((item) => {
      if (item === null) {
        return false;
      } else {
        return parseInt(item.id) === attachment.id;
      }
    });
    if (index !== -1) {
      var fileUploader = temp[index];
      fileUploader.click();
    }
  };

  onChangeFile =  async(event) => {
    event.stopPropagation();
    event.preventDefault();
    var file = event.target.files[0];
    let id_ = parseInt(event.target.id);

    if (file !== undefined && file !== null) {
      var index_ = -1;
      index_ = this.props.attachments.findIndex((item) => {
        return item.id === id_;
      });

      if (index_ === -1) {
        if (process.env.NODE_ENV === "development") {
          console.log("[ERROR] Can't find the event.target.id:", id_);
        }
      } else {
        let url = `/api/Document/UpdateAttachment`;
        let payload = { id: id_ };

        let response =  await postAttachmentData(payload, file, url);

        if (response.ok) {
          let temp = this.props.attachments;
          temp[index_].name = file.name;
          temp[index_].contentType = file.type;
          this.setState(
            {
              openNotifiction: true,
            },
            this.updateParentAttachments(temp)
          );
        } else {
          let json = response.json();
          this.setState({
            openNotifiction: true,
            error: true,
            errorMessage: json.error,
          });
        }
      }
    }
  };

  generateDownloadLinks(attachments) {
    let links = attachments.map((attachment, index) => {
      return (
        <a
        id={"link" + index}
          key={index}
          style={{ display: "none" }}
          href={attachment.attachmentFile}
          download={attachment.name}
        >
          <button
            id={`download-attachment-button${attachment.id}`}
            type="submit"
          >
            {this.t("download")}
          </button>
        </a>
      );
    });
    return links;
  }

  generateInputFiles(attachments) {
    let links = attachments.map((attachment, index) => {
      return (
        <input
          type="file"
          className={"onChangeFile" + attachment.id}
          id={attachment.id}
          key={attachment.id}
          ref={(ref) => {
            this.inputRefs.add(ref);
          }}
          style={{ display: "none" }}
          onChange={this.onChangeFile}
        />
      );
    });
    return links;
  }

  setOpenNotification = (value) => {
    this.setState({
      openNotifiction: value,
      error: false,
      errorMessage: [],
    });
  };

  render() {
    return (
      <div>
        {this.generateDownloadLinks(this.props.attachments)}
        {this.generateInputFiles(this.props.attachments)}
        <Notification
        id="notifications"
          open={this.state.openNotifiction} // must be a "state" to show notification message [true or false]
          setOpen={this.setOpenNotification} // this function changes the state "open" value.
          error={this.state.error} // if error message put [true] else for success put [false]
          errorMessage={this.state.errorMessage} // the recived error message (array of string) from server.
        />
        <MaterialTable
          title={this.t("attachment_table")}
          icons={tableIcons}
          columns={[
            { title: this.t("attachment_id"), field: "id", hidden: true },
            { title: this.t("caption"), field: "caption" },
            { title: this.t("attachment_name"), field: "name" },
            { title: this.t("content_type"), field: "contentType" },
            {
              title: this.t("attachment_file"),
              field: "attachmentFile",
              hidden: true,
            },
            { title: this.t("added_date"), field: "addedDate" },
          ]}
          data={this.props.attachments}
          actions={[
            {
              icon: DeleteIcon,
              tooltip: this.t("delete"),
              onClick: (event, rowData) => {
                this.deleteAttachment(rowData);
              },
            },
            {
              icon: GetAppIcon,
              tooltip: this.t("delete"),
              onClick: (event, rowData) => {
                document
                  .getElementById(`download-attachment-button${rowData.id}`)
                  .click();
              },
            },
            {
              icon: UpdateIcon,
              tooltip: this.t("update"),
              onClick: (event, rowData) => {
                this.updateAttachmentcOnClickHandler(rowData);
              },
            },
          ]}
          options={{
            actionsColumnIndex: -1,
            exportButton: true,
          }}
          localization={{
            pagination: {
              labelRowsSelect: this.t("rows"),
              firstAriaLabel: this.t("first_page"),
              firstTooltip: this.t("first_page"),
              previousAriaLabel: this.t("previous_page"),
              previousTooltip: this.t("previous_page"),
              nextAriaLabel: this.t("next_page"),
              nextTooltip: this.t("next_page"),
              lastAriaLabel: this.t("last_page"),
              lastTooltip: this.t("last_page"),
            },
            toolbar: {
              nRowsSelected: "{0}" + this.t("selected_row"),
              searchTooltip: this.t("search"),
              searchPlaceholder: this.t("search"),
              exportTitle: this.t("export_as_csv"),
            },
            header: {
              actions: this.t("actions"),
            },
            body: {
              emptyDataSourceMessage: this.t("no_records"),
            },
          }}
        />
      </div>
    );
  }
}

export default withTranslation()(AttachmentTable);
