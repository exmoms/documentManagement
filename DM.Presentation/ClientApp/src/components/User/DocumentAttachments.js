/**
 * @file Contains ProcessAttachments component implementaion
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @version 1.0
 */

import React, { Component } from "react";
import AttachmentTable from "./AttachmentTable";
import { Notification } from "../Admin/Notifications";

import { postAttachmentData } from "../../api/PostData";
import CREATE_DOC_STYLES from "./Styles";
import { withStyles } from "@material-ui/core/styles";
import { fetchData } from "../../api/FetchData";
import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
// import DialogTitle from "@material-ui/core/DialogTitle";
import DialogContent from "@material-ui/core/DialogContent";
import DialogActions from "@material-ui/core/DialogActions";
import { withTranslation } from "react-i18next";

const styles = CREATE_DOC_STYLES;

class DocumentAttachments extends Component {
  constructor(props) {
    super(props);

    this.state = {
      documentId: -1,
      attachments: [],
    };
    this.fileInputRef = React.createRef();
  }

  t = this.props["t"];

  fetchDocumentAttachments = async (documentId) => {
    let attachments = [];
    let url = `/api/Document/GetAttachmentsByDocumentId?docId=${documentId}`;
    await fetchData(url).then((data) => {
      if (data.length !== 0) {
        attachments = data.map((attachment) => {
          return {
            id: attachment.id,
            name: attachment.name,
            contentType: attachment.contentType,
            attachmentFile: attachment.attachmentFile,
            compoundModelId:
              attachment.compoundModelId !== undefined
                ? attachment.compoundModelId
                : null,
            caption: attachment.caption,
            addedDate: attachment.addedDate,
          };
        });
      }
    });
    return attachments;
  };

  async componentDidMount() {
    if (this.props.documentId !== this.state.documentId) {
      let attachments = await this.fetchDocumentAttachments(
        this.props.documentId
      );

      this.setState({
        documentId: this.props.documentId,
        attachments: attachments,
      });
    }
  }

  async componentDidUpdate() {
    if (this.props.documentId !== this.state.documentId) {
      let attachments = await this.fetchDocumentAttachments(
        this.props.documentId
      );

      this.setState({
        documentId: this.props.documentId,
        attachments: attachments,
      });
    }
  }

  addNewAttachmentOnClickHandler = () => {
    var fileUploader = this.fileInputRef;
    fileUploader.click();
  };

  onChangeFile = async (event) => {
    event.stopPropagation();
    event.preventDefault();
    var file = event.target.files[0];
    let id_ = this.props.documentId;

    if (file !== undefined && file !== null) {
      let url = "/api/Document/AddNewAttachmentToDocument";
      let payload = { DocumentId: id_ };
      let upload = false;

      let response = await postAttachmentData(payload, file, url);

      if (response.ok) {
        upload = true;
      } else {
        let json = await response.json();
        this.setState({
          openNotifiction: true,
          error: true,
          errorMessage: json.error,
        });
      }

      if (upload) {
        let attachments = await this.fetchDocumentAttachments(id_);
        this.setState({
          attachments: attachments,
          openNotifiction: true,
        });
      }
    }
  };

  handleCloseDialog = () => {
    this.props.handleClose();
  };

  updateAttachments = (attachments) => {
    this.setState({
      attachments: attachments,
    });
  };

  setOpenNotification = (value) => {
    this.setState({
      openNotifiction: value,
      error: false,
      errorMessage: [],
    });
  };

  render() {
    let { documentId } = this.props;
    return (
      <div>
        <input
          type="file"
          className={"onChangeFile" + documentId}
          id={documentId}
          key={documentId}
          //   required={attachment.isRequiered}
          ref={(ref) => {
            this.fileInputRef = ref;
          }}
          style={{ display: "none" }}
          onChange={this.onChangeFile}
        />

        <Dialog
          position="centerCenter"
          open={this.props.show}
          onClose={() => this.handleCloseDialog()}
          aria-labelledby="form-dialog-title"
        >
          {/* <DialogTitle id="form-dialog-title">Attachments Dialog</DialogTitle> */}
          <DialogContent>
            <Notification
              open={this.state.openNotifiction} // must be a "state" to show notification message [true or false]
              setOpen={this.setOpenNotification} // this function changes the state "open" value.
              error={this.state.error} // if error message put [true] else for success put [false]
              errorMessage={this.state.errorMessage} // the recived error message (array of string) from server.
            />
            <AttachmentTable
              attachments={this.state.attachments}
              updateAttachments={this.updateAttachments}
            />
          </DialogContent>
          <DialogActions>
            <Button
              onClick={() => this.handleCloseDialog()}
              id="toggleShowButton"
              color="primary"
            >
              {this.t("button_close")}
            </Button>
            <Button
              onClick={() => {
                this.addNewAttachmentOnClickHandler();
              }}
              id="toggleShowButton"
              color="primary"
            >
              {this.t("add_attachment")}
            </Button>
          </DialogActions>
        </Dialog>
      </div>
    );
  }
}

export default withStyles(styles)(withTranslation()(DocumentAttachments));
