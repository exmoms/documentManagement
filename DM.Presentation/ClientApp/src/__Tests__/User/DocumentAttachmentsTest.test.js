import React from "react";
import ReactDom from "react-dom";
import DocumentAttachments from "../../components/User/DocumentAttachments";
import AttachmentTable from "../../components/User/AttachmentTable";
import { cleanup } from "@testing-library/react";
import { mount } from "enzyme";
import Form from "react-jsonschema-form";
import { shallow } from "enzyme";
import { act } from "react-dom/test-utils";
import Dialog from "@material-ui/core/Dialog";
import * as FetchMock from "../../api/FetchData";
import * as PostDataMock from "../../api/PostData";
import Button from "@material-ui/core/Button";
afterEach(cleanup);
describe("DocumentAttachments", () => {
    afterEach(() => {
        jest.clearAllMocks();
      });
      test("handleCloseDialog", () => {
        jest.clearAllMocks();
        let documentIdMock = -1;
        let show  = true;
        const handleClose = jest.fn();
        const mockSuccessResponse = [];
        const mockJsonPromise = Promise.resolve(mockSuccessResponse); 
        const PostMock =  jest.spyOn(PostDataMock, "postAttachmentData").mockImplementation(() => mockJsonPromise);
      const mockSuccessResponse2 = [{id : 1 , name:"attachment A" , contentType: "type" , attachmentFile:"attachmentFile" , compoundModelId : 2 ,caption :"caption" , addedDate:'29/4/2020'}];
          const mockJsonPromise2 = Promise.resolve(mockSuccessResponse2); 
          const fetchMock =  jest.spyOn(FetchMock, "fetchData").mockImplementation(() => mockJsonPromise2);
      let wrapper;
      const stopPropagationMock = jest.fn();
      return mockJsonPromise2.then(() => {  
       wrapper = mount(
            <DocumentAttachments
            documentId={documentIdMock}
          show={show}
          handleClose={handleClose}
            />
        );
        act(() => {
          wrapper
            .find(Dialog)
            .at(0)
            .props()
            .onClose();
        });
      }).then(() => {
        wrapper.update();
        wrapper.render();
        expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
        expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual([]);
        
        expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
        expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual([
           ]);
        expect(handleClose).toHaveBeenCalledTimes(1);
      });
        });
      
        test("updateAttachments", () => {
          jest.clearAllMocks();
          let documentIdMock = -1;
          let show  = true;
          const handleClose = jest.fn();
          const mockSuccessResponse = [];
        const mockJsonPromise = Promise.resolve(mockSuccessResponse); 
        const PostMock =  jest.spyOn(PostDataMock, "postAttachmentData").mockImplementation(() => mockJsonPromise);
          const mockSuccessResponse2 = [{id : 1 , name:"attachment A" , contentType: "type" , attachmentFile:"attachmentFile" , compoundModelId : 2 ,caption :"caption" , addedDate:'29/4/2020'}];
          const mockJsonPromise2 = Promise.resolve(mockSuccessResponse2); 
          const fetchMock =  jest.spyOn(FetchMock, "fetchData").mockImplementation(() => mockJsonPromise2);
          let wrapper;
          let attachments =[{id : 1 , attchmentName: "attachment1"} , {id : 2 , attchmentName: "attachment2"}]
          return mockJsonPromise2.then(() => {  
           wrapper = mount(
              <DocumentAttachments
              documentId={documentIdMock}
            show={show}
            handleClose={handleClose}
              />
          );
          act(() => {
              wrapper
                .find(AttachmentTable)
                .at(0)
                .props()
                .updateAttachments(attachments);
            });
        }).then(() => {
            wrapper.update();
          wrapper.render();
          expect(wrapper.find('.onChangeFile1').at(0).props().id).toBe(1);
          expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
          expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual(attachments);
        });
      });
        test("addNewAttachmentOnClickHandler", () => {
          jest.clearAllMocks();
          let documentIdMock = -1;
          let show  = true;
          const handleClose = jest.fn();
          const mockSuccessResponse = [];
        const mockJsonPromise = Promise.resolve(mockSuccessResponse); 
        const PostMock =  jest.spyOn(PostDataMock, "postAttachmentData").mockImplementation(() => mockJsonPromise);
          const mockSuccessResponse2 = [{id : 1 , name:"attachment A" , contentType: "type" , attachmentFile:"attachmentFile" , compoundModelId : 2 ,caption :"caption" , addedDate:'29/4/2020'}];
        const mockJsonPromise2 = Promise.resolve(mockSuccessResponse2); 
        const fetchMock =  jest.spyOn(FetchMock, "fetchData").mockImplementation(() => mockJsonPromise2);
        let wrapper;
        const clickInputSpy = jest.spyOn(HTMLInputElement.prototype, 'click');
        return mockJsonPromise2.then(() => {  
           wrapper = mount(
              <DocumentAttachments
              documentId={documentIdMock}
            show={show}
            handleClose={handleClose}
              />
          );
          
        }).then(() => {
          wrapper.update();
          wrapper.render();
          act(() => {
              wrapper
                .find(Button)
                .at(1)
                .props()
                .onClick();
            });
          expect(clickInputSpy).toHaveBeenCalledTimes(1);
          expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
          expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual([]);
        });
        });
  test("renders without crashing", () => {
    const div = document.createElement("div");
    let documentId = 1;
    let show  = true;
    const handleClose = jest.fn();
    ReactDom.render(
      <DocumentAttachments
      documentId={documentId}
      show={show}
      handleClose={handleClose}
      />,
      div
    );
  });

  it("matches the snapshot", () => {
    let documentId = 1;
    let show  = true;
    const handleClose = jest.fn();
    const wrapper = mount(
        <DocumentAttachments
        documentId={documentId}
      show={show}
      handleClose={handleClose}
        />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("DocumentAttachments intializations", () => {
    let documentIdMock = 1;
    let show  = true;
    const handleClose = jest.fn();
    const wrapper = mount(
        <DocumentAttachments
        documentId={documentIdMock}
      show={show}
      handleClose={handleClose}
        />
    );
    expect(wrapper.find('.onChangeFile1').at(0).props().id).toBe(1);
    expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
    expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual([]);
  });

//   test("componentDidUpdate functionalities", () => {
//     let documentIdMock = 1;
//     let show  = true;
//     const handleClose = jest.fn();
//     const mockSuccessResponse = [{id : 1 , name:"attachment A" , contentType: "type" , attachmentFile:"attachmentFile" , compoundModelId : 2 ,caption :"caption" , addedDate:'29/4/2020'}];
//   const mockJsonPromise = Promise.resolve(mockSuccessResponse); 
//   const fetchMock =  jest.spyOn(FetchMock, "fetchData").mockImplementation(() => mockJsonPromise);
//   let wrapper;
//   return mockJsonPromise.then(() => {  
//    wrapper = mount(
//         <DocumentAttachments
//         documentId={documentIdMock}
//       show={show}
//       handleClose={handleClose}
//         />
//     );
//     expect(wrapper.find('.onChangeFile1').props().id).toBe(1);
//     expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
//     expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual([]);
//     //calling update
//     wrapper.setProps({documentId:2})
//   }).then(() => {
//     wrapper.update();
//     wrapper.render();
//     expect(fetchMock).toHaveBeenCalledTimes(2);
//     expect(fetchMock.mock.calls[0][0]).toBe("/api/Document/GetAttachmentsByDocumentId?docId=1");
//     expect(wrapper.find('.onChangeFile2').props().id).toBe(2);
//     expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
//     expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual(mockSuccessResponse);
// });
  
// });

  test("componentDidUpdate functionalities when length is 0", () => {
    let documentIdMock = 1;
    let show  = true;
    const handleClose = jest.fn();
    const mockSuccessResponse = [];
  const mockJsonPromise = Promise.resolve(mockSuccessResponse); 
  const fetchMock =  jest.spyOn(FetchMock, "fetchData").mockImplementation(() => mockJsonPromise);
  let wrapper;
  return mockJsonPromise.then(() => {  
   wrapper = mount(
        <DocumentAttachments
        documentId={documentIdMock}
      show={show}
      handleClose={handleClose}
        />
    );
    expect(wrapper.find('.onChangeFile1').props().id).toBe(1);
    expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
    expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual([]);
    //calling update
    wrapper.setProps({documentId:2})
  }).then(() => {
    wrapper.update();
    wrapper.render();
    expect(fetchMock).toHaveBeenCalledTimes(2);
    expect(fetchMock.mock.calls[0][0]).toBe("/api/Document/GetAttachmentsByDocumentId?docId=1");
    expect(wrapper.find('.onChangeFile2').props().id).toBe(2);
    expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
    expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual([]);
});
});
 
test("onChangeFile functionalities", () => {
    let documentIdMock = 1;
    let show  = true;
    const handleClose = jest.fn();
    const mockSuccessResponse = {ok:true  , json:()=>{ return {error:["errorMessage"]}}};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse); 
  const PostMock =  jest.spyOn(PostDataMock, "postAttachmentData").mockImplementation(() => mockJsonPromise);
  const fetchMock =  jest.spyOn(FetchMock, "fetchData").mockImplementation(() => Promise.resolve([]));
  let wrapper;
  const stopPropagationMock = jest.fn();
    const preventDefaultMock = jest.fn();
  return mockJsonPromise.then(() => {  
   wrapper = mount(
        <DocumentAttachments
        documentId={documentIdMock}
      show={show}
      handleClose={handleClose}
        />
    );
    expect(wrapper.find('.onChangeFile1').props().id).toBe(1);
    expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
    expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual([]);
    act(() => {
        wrapper
          .find(".onChangeFile1")
          .at(0)
          .props()
          .onChange({
            target: { id: 1, files: [{ name: "file1" }] },
            stopPropagation: stopPropagationMock,
            preventDefault: preventDefaultMock,
          });
      });
  }).then(() => {
    wrapper.update();
    wrapper.render();
    expect(stopPropagationMock).toHaveBeenCalledTimes(1);
    expect(preventDefaultMock).toHaveBeenCalledTimes(1);
    expect(PostMock).toHaveBeenCalledTimes(1);
    expect(PostMock.mock.calls[0][0]).toStrictEqual({DocumentId: 1 });
    expect(PostMock.mock.calls[0][1]).toStrictEqual({ name: "file1" });
    expect(PostMock.mock.calls[0][2]).toStrictEqual("/api/Document/AddNewAttachmentToDocument");
    expect(wrapper.find('.onChangeFile1').props().id).toBe(1);
    expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
    expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual([]);
    expect(fetchMock).toHaveBeenCalledTimes(2);
    expect(fetchMock.mock.calls[0][0]).toBe("/api/Document/GetAttachmentsByDocumentId?docId=1");
});
});

test("onChangeFile functionalities when upload is false", () => {
    let documentIdMock = 1;
    let show  = true;
    const handleClose = jest.fn();
    const mockSuccessResponse = {ok:false  , json:()=>{ return {error:["errorMessage"]}}};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse); 
  const PostMock =  jest.spyOn(PostDataMock, "postAttachmentData").mockImplementation(() => mockJsonPromise);
  const mockSuccessResponse2 = [{id : 1 , name:"attachment A" , contentType: "type" , attachmentFile:"attachmentFile" , compoundModelId : 2 ,caption :"caption" , addedDate:'29/4/2020'}];
  const mockJsonPromise2 = Promise.resolve(mockSuccessResponse2); 
  const fetchMock =  jest.spyOn(FetchMock, "fetchData").mockImplementation(() => mockJsonPromise2);
  let wrapper;
  const stopPropagationMock = jest.fn();
    const preventDefaultMock = jest.fn();
  return mockJsonPromise2.then(() => {  
   wrapper = mount(
        <DocumentAttachments
        documentId={documentIdMock}
      show={show}
      handleClose={handleClose}
        />
    );
  }).then(() => {
    wrapper.update();
    wrapper.render();
    expect(wrapper.find('.onChangeFile1').props().id).toBe(1);
    expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
    expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual([]);
    return mockJsonPromise.then(() => { 
    act(() => {
        wrapper
          .find(".onChangeFile1")
          .at(0)
          .props()
          .onChange({
            target: { id: 1, files: [{ name: "file1" }] },
            stopPropagation: stopPropagationMock,
            preventDefault: preventDefaultMock,
          });
      });
  }).then(() => {
    wrapper.update();
    wrapper.render();
    expect(stopPropagationMock).toHaveBeenCalledTimes(1);
    expect(preventDefaultMock).toHaveBeenCalledTimes(1);
    expect(PostMock).toHaveBeenCalledTimes(1);
    expect(PostMock.mock.calls[0][0]).toStrictEqual({DocumentId: 1 });
    expect(PostMock.mock.calls[0][1]).toStrictEqual({ name: "file1" });
    expect(PostMock.mock.calls[0][2]).toStrictEqual("/api/Document/AddNewAttachmentToDocument");
    expect(wrapper.find('.onChangeFile1').at(0).props().id).toBe(1);
    expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
    expect(wrapper.find(AttachmentTable).at(0).props().attachments).toStrictEqual([
          {
         "addedDate": "29/4/2020",
          "attachmentFile": "attachmentFile",
           "caption": "caption",
           "compoundModelId": 2,
           "contentType": "type",
           "id": 1,
           "name": "attachment A",
           "tableData":  {
             "id": 0,
           },
         },
       ]);
    expect(fetchMock).toHaveBeenCalledTimes(1);
});
});
});
});