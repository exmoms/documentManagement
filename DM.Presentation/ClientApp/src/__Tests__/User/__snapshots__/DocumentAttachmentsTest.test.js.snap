// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentAttachments matches the snapshot 1`] = `ReactWrapper {}`;
